### PR TITLE
Normalize rmw_time_t according to DDS spec

### PR DESCRIPTION
**Pull request by asorbini**
This PR updates the implementation of `rmw_dds_common::clamp_rmw_time_to_dds_time()` to better comply with the DDS specification, which mandates that time values be normalized to have the "nanoseconds" component always < 1 second.

Quoting the relevant bit from the spec ([v1.4](), section 2.3.2):
> The two types used to represent time: Duration_t and Time_t are mapped into structures that contain fields for the second and the nanosecond parts. These types are further constrained to always use a ‘normalized’ representation for the time, that is, the nanosec field must verify 0 <= nanosec < 1000000000.

The updates limit the values produced by `rmw_dds_common::clamp_rmw_time_to_dds_time() to a maximum of {`INT_MAX` seconds, `10^9 - 1` nanoseconds}.

I came up with these changes while working on [rmw_connextdds#21](https://github.com/ros2/rmw_connextdds/issues/21), and even though `rmw_connextdds` will not directly use this function for `WaitSet::wait()` going forward, I'm still submitting the PR because the function will be used again in the future for other purposes (e,g, see [rmw_connextdds#20](https://github.com/ros2/rmw_connextdds/issues/20)), and this might be a slightly "safer" conversion than just truncating at `[U]INT_MAX`, because it guarantees that any compliant DDS implementation will be able to represent the output.

The implementation intentionally doesn't handle constant `RMW_DURATION_INFINITE`, because it expects the caller to handle that case explicitly. If desired, the function could be extended to handle this value (and maybe produce in output `{ 0x7FFFFFFF, 0x7FFFFFFF }` which the DDS spec defines as `DDS_DURATION_INFINITE`).

While reviewing the value of `RMW_DURATION_INFINITE` I have also been brought to wonder why it wasn't defined to be `{ UINT64_MAX, UINT64_MAX }` and it instead uses a special canary value (`INT64_MAX` split between the two fields). It's not so much a problem when it comes to DDS (since the value is beyond what can be represented), but API-wise, I would have expected this special value to fall on some boundary of the "natural domain" of the type, rather than somewhere in the middle.

### CODE REVIEW

**Comment by emersonknapp** [comment]
> API-wise, I would have expected this special value to fall on some boundary of the "natural domain" of the type, rather than somewhere in the middle.

For context, the reasoning is that most of the ros2 codebase uses an `int64_t nanoseconds` representation to represent durations. Only the RMW interface uses a split seconds+nanoseconds struct. This was considered good because it's the maximum duration, say, rclcpp can represent. Unfortunately the downside is that it looks a little awkward in DDS.

File: rmw_dds_common/src/time_utils.cpp
Comment (hidmic) [inline rmw_dds_common/src/time_utils.cpp:32]: @asorbini nit: 
```suggestion
  constexpr uint64_t sec_to_ns = 1000000000ULL;
```
will likely get you what you want.

File: rmw_dds_common/src/time_utils.cpp
Comment (hidmic) [inline rmw_dds_common/src/time_utils.cpp:35]: @asorbini nit^N:
```suggestion
  bool overflow_nsec = false;
  bool overflow_sec = false;
```
is more verbose but easier to read IMHO.

File: rmw_dds_common/src/time_utils.cpp
Comment (hidmic) [inline rmw_dds_common/src/time_utils.cpp:49]: @asorbini why checking for `ns_sec_adjust == 0`? If `t.sec` overflows, we want `t.nsec` saturated unconditionally, no?

File: rmw_dds_common/test/test_time_utils.cpp
Comment (hidmic) [inline rmw_dds_common/test/test_time_utils.cpp:51]: @asorbini consider adding a test case for when both seconds and nanoseconds are slightly above the limit.

File: rmw_dds_common/src/time_utils.cpp
Comment (asorbini) [reply rmw_dds_common/src/time_utils.cpp]: Applied in [1fb9751](https://github.com/ros2/rmw_dds_common/pull/48/commits/1fb97517ff4ed002e510664b6906deae39c4322f).

File: rmw_dds_common/src/time_utils.cpp
Comment (asorbini) [reply rmw_dds_common/src/time_utils.cpp]: Applied in [1fb9751](https://github.com/ros2/rmw_dds_common/pull/48/commits/1fb97517ff4ed002e510664b6906deae39c4322f).

File: rmw_dds_common/test/test_time_utils.cpp
Comment (asorbini) [reply rmw_dds_common/test/test_time_utils.cpp:51]: Test added in [1fb9751](https://github.com/ros2/rmw_dds_common/pull/48/commits/1fb97517ff4ed002e510664b6906deae39c4322f).

File: rmw_dds_common/src/time_utils.cpp
Comment (asorbini) [reply rmw_dds_common/src/time_utils.cpp]: Yes, you are right. By the time we hit that block, if either of those `overflow` variable is true, then we should saturate the nanosecond component.

Applied in [1fb9751](https://github.com/ros2/rmw_dds_common/pull/48/commits/1fb97517ff4ed002e510664b6906deae39c4322f).

File: rmw_dds_common/src/time_utils.cpp
Comment (hidmic) [inline rmw_dds_common/src/time_utils.cpp:60]: @asorbini nit: would it make sense to merge the last two `if` statements? Their conditions are identical.

**Review by hidmic — APPROVED** [review]
LGTM pending green CI

File: rmw_dds_common/src/time_utils.cpp
Comment (asorbini) [reply rmw_dds_common/src/time_utils.cpp:55]: Good point, fixed in d38c995

**Comment by hidmic** [comment]
CI up to `rmw_dds_common` and `test_rmw_implementation`:

* Linux [Build Status](http://ci.ros2.org/job/ci_linux/14632/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/9386/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/12297/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/14777/)

**Comment by hidmic** [comment]
This could be backported to Galactic. It can probably wait till patch release 1.

**Comment by hidmic** [comment]
Alright. Moving forward. Thanks for the fix @asorbini.

**Pull request merged**
